include/uk/arch/spinlock.h

```diff
@@ -21,20 +23,62 @@
  * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
  * SUCH DAMAGE.
  */
-/* Taken from Mini-OS */
 
 #ifndef __UKARCH_SPINLOCK_H__
 #define __UKARCH_SPINLOCK_H__
 
+#include <uk/config.h>
+#include <stdint.h>
+
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-typedef struct {} spinlock_t;
+#ifdef CONFIG_UKPLAT_LCPU_MULTICORE
+/** spinlock struct */
+typedef struct {
+	volatile uint32_t lock;
+} __spinlock_t;
+
+/** section attribute cannot be used with pointers */
+typedef __spinlock_t *spinlock_ptr_t;
+
+/* spinlock type for declaration */
+#define spinlock_t __attribute__((section(".slocks"))) __spinlock_t
+
+/**
+ * Initialize a spinlock object (put in unlocked state)
+ * @param [in/out] lock Pointer to spinlock object
+ */
+void ukarch_spin_lock_init(spinlock_ptr_t spinlock);
```

Comment:
According to `__atomic` we would rename the functions as

```C
void __spinlock_init(__spinlock *lock);
void __spinlock_lock(__spinlock *lock);
void __spinlock_unlock(__spinlock *lock);
int __spinlock_is_locked(__spinlock *lock);
int __spinlock_trylock(__spinlock *lock);
```

Having the word `spinlock` in it also makes it more intuitive to `grep` for spinlock functions.